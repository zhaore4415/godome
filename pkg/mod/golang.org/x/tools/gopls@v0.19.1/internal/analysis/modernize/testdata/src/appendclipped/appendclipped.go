package appendclipped

import (
	"os"
	"slices"
)

type (
	Bytes  []byte
	Bytes2 []byte
)

func _(s, other []string) {
	print(append([]string{}, s...))              // want "Replace append with slices.Clone"
	print(append([]string(nil), s...))           // want "Replace append with slices.Clone"
	print(append(Bytes(nil), Bytes{1, 2, 3}...)) // want "Replace append with slices.Clone"
	print(append(other[:0:0], s...))             // want "Replace append with slices.Clone"
	print(append(other[:0:0], os.Environ()...))  // want "Redundant clone of os.Environ()"
	print(append(other[:0], s...))               // nope: intent may be to mutate other

	print(append(append(append([]string{}, s...), other...), other...))             // want "Replace append with slices.Concat"
	print(append(append(append([]string(nil), s...), other...), other...))          // want "Replace append with slices.Concat"
	print(append(append(Bytes(nil), Bytes{1, 2, 3}...), Bytes{4, 5, 6}...))         // want "Replace append with slices.Concat"
	print(append(append(append(other[:0:0], s...), other...), other...))            // want "Replace append with slices.Concat"
	print(append(append(append(other[:0:0], os.Environ()...), other...), other...)) // want "Replace append with slices.Concat"
	print(append(append(other[:len(other):len(other)], s...), other...))            // want "Replace append with slices.Concat"
	print(append(append(slices.Clip(other), s...), other...))                       // want "Replace append with slices.Concat"
	print(append(append(append(other[:0], s...), other...), other...))              // nope: intent may be to mutate other
}

var (
	_ Bytes  = append(Bytes(nil), []byte(nil)...) // nope: correct fix requires Clone[Bytes] (#73661)
	_ Bytes  = append([]byte(nil), Bytes(nil)...) // nope: correct fix requires Clone[Bytes] (#73661)
	_ Bytes2 = append([]byte(nil), Bytes(nil)...) // nope: correct fix requires Clone[Bytes2] (#73661)
)
