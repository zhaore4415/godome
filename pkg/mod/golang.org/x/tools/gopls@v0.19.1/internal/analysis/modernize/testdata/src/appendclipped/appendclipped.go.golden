package appendclipped

import (
	"os"
	"slices"
)

type (
	Bytes	[]byte
	Bytes2  []byte
)

func _(s, other []string) {
	print(slices.Clone(s))              // want "Replace append with slices.Clone"
	print(slices.Clone(s))              // want "Replace append with slices.Clone"
	print(slices.Clone(Bytes{1, 2, 3})) // want "Replace append with slices.Clone"
	print(slices.Clone(s))              // want "Replace append with slices.Clone"
	print(os.Environ())                 // want "Redundant clone of os.Environ()"
	print(append(other[:0], s...))      // nope: intent may be to mutate other

	print(slices.Concat(s, other, other))                              // want "Replace append with slices.Concat"
	print(slices.Concat(s, other, other))                              // want "Replace append with slices.Concat"
	print(slices.Concat(Bytes{1, 2, 3}, Bytes{4, 5, 6}))               // want "Replace append with slices.Concat"
	print(slices.Concat(s, other, other))                              // want "Replace append with slices.Concat"
	print(slices.Concat(os.Environ(), other, other))                   // want "Replace append with slices.Concat"
	print(slices.Concat(other, s, other))                              // want "Replace append with slices.Concat"
	print(slices.Concat(other, s, other))                              // want "Replace append with slices.Concat"
	print(append(append(append(other[:0], s...), other...), other...)) // nope: intent may be to mutate other
}

var (
	_ Bytes  = append(Bytes(nil), []byte(nil)...) // nope: correct fix requires Clone[Bytes] (#73661)
	_ Bytes  = append([]byte(nil), Bytes(nil)...) // nope: correct fix requires Clone[Bytes] (#73661)
	_ Bytes2 = append([]byte(nil), Bytes(nil)...) // nope: correct fix requires Clone[Bytes2] (#73661)
)
